Implement `field-sizing: content` minimum width for text caret

lukewarlow opened the pull request:
#### c2e64ba96fbed90353cfdea4b0c20a855a6c40bb
<pre>
Implement `field-sizing: content` minimum width for text caret
<a href="https://bugs.webkit.org/show_bug.cgi?id=269128">https://bugs.webkit.org/show_bug.cgi?id=269128</a>

Reviewed by Darin Adler.

Updates text control computed style to include a minimum sizing so the caret shows.

* LayoutTests/fast/forms/text/text-field-sizing-caret-expected.txt: Added.
* LayoutTests/fast/forms/text/text-field-sizing-caret.html: Added.
* LayoutTests/fast/forms/textarea/textarea-field-sizing-caret-expected.txt: Added.
* LayoutTests/fast/forms/textarea/textarea-field-sizing-caret.html: Added.
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const):

Canonical link: <a href="https://commits.webkit.org/279115@main">https://commits.webkit.org/279115@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b272fbe30ecf4d20b1685066fc8f9196c8912c8d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52227 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31558 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55501 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2950 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54531 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2648 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42504 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1896 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45081 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23587 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2355 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1109 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48361 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2503 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57097 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2556 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28586 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45199 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11474 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29498 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28330 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -907,6 +907,13 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
 
+    if (parentStyle.fieldSizing() == FieldSizing::Content) {
+        if (textBlockStyle.isHorizontalWritingMode())
+            textBlockStyle.setMinWidth(Length { 1, LengthType::Fixed });
```
I think this needs to use `caretWidth()` rather than 1, since the width can be 2 on Cocoa platforms.

nt1m replied:
Does this override whatever min-width the web developer has set? If the developer set `min-width: 200px;`, I expect that to be honored. It might be better to change the intrinsic size to account for the caret instead?

lukewarlow replied:
The developer can't set a min width as this is for the inner element of the input rather than the input itself. I will make sure it works as expected though. This is how chromium have implemented this fwiw

nt1m replied:
Ah right, targeting the inner element might be more reasonable then.

pxlcoder commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -907,6 +907,13 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
 
+    if (parentStyle.fieldSizing() == FieldSizing::Content) {
+        if (textBlockStyle.isHorizontalWritingMode())
+            textBlockStyle.setMinWidth(Length { 1, LengthType::Fixed });
```
`setLogicalMinWidth`, and you can remove the `if`/`else`.

darinadler commented on LayoutTests/fast/forms/textarea/textarea-field-sizing-caret.html:
```diff
@@ -0,0 +1,22 @@
+}, 'The width of an empty textarea should be same as the caret width');
+test(() => {
+  assert_equals(document.querySelectorAll('textarea')[1].offsetHeight, finalCaretRect.width);
+}, 'The height of an empty textarea should be same as the caret height in a vertical wirting-mode');
```
This has a typo in the word "writing".

darinadler commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -907,6 +907,9 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
         }
     }
 
+    if (parentStyle.fieldSizing() == FieldSizing::Content)
+        textBlockStyle.setLogicalMinWidth(Length { caretWidth(), LengthType::Fixed });
```
Is there a way for this to be tested in Web Platform Tests instead of just in WebKit-specific tests? It seems like the kind of thing we want to ensure is consistent across browsers.

lukewarlow replied:
I'm not familiar with anyway to get the size of a caret using exposed web platform APIs.

lukewarlow replied:
I imagine if there was then the internals function to get the caret bounding box wouldn't be needed?

pxlcoder replied:
Right, the caret width differs (at the least) on Cocoa platforms, so any test would need to have a way to directly/indirectly get the width.

Interestingly, this patch might be the first place to (indirectly) expose caret width to the web?

Could we still write a WPT that asserts a greater-than-or-equal-to width, in addition to keeping this WebKit-specific test?

pxlcoder replied:
It looks like there is specified API to get information about the caret, `CaretPosition.getClientRect()`, but it's only implemented in Firefox.

https://drafts.csswg.org/cssom-view/#dom-caretposition-getclientrect

lukewarlow replied:
One other reason why we might not want to add a WPT for this, is that the spec only says that agents "may" take the caret width into account. It doesn't require it? https://html.spec.whatwg.org/#the-textarea-element-2

darinadler replied:
Open ended specifications like that are not great for interoperability!

Reflecting on it, I think space for the caret is needed alongside the text, so the **minimum** size isn’t all that’s needed, if someone is trying to fit a textarea to the size of the text that will be inside it.

lukewarlow replied:
> Open ended specifications like that are not great for interoperability!

I agree fwiw and I don't think there's a reason a UA shouldn't account for the caret width as a minimum.

> Reflecting on it, I think space for the caret is needed alongside the text

The only situation where there appears to be an issue with hidden carets is when the content size is 0px (an empty input or textarea). When there's text in the control the caret shows fine from my testing. I'd also personally prefer to match Chromium as close as possible and then raise any spec issues necessary to reach improvements if they're needed.

lukewarlow replied:
Am I okay to get this PR merged or would you prefer me to add a minimum caret width test to WPT?

darinadler replied:
Oh that was never intended to block merging this change in. If the concerns about adding a way to detect the width to the web platform and fingerprinting are not serious then I don't know of any blocker.

lukewarlow replied:
Okay I'll get this merged, as it's behind a flag either way. But given the above mentioned API about getting caret information + the fact I don't think this width is customisable (so it exposes nothing more than the UA string already does), I don't think there's any novel fingerprinting surface here.

lukewarlow committed:
Implement `field-sizing: content` minimum width for text caret
https://bugs.webkit.org/show_bug.cgi?id=269128

Reviewed by Darin Adler.

Updates text control computed style to include a minimum sizing so the caret shows.

* LayoutTests/fast/forms/text/text-field-sizing-caret-expected.txt: Added.
* LayoutTests/fast/forms/text/text-field-sizing-caret.html: Added.
* LayoutTests/fast/forms/textarea/textarea-field-sizing-caret-expected.txt: Added.
* LayoutTests/fast/forms/textarea/textarea-field-sizing-caret.html: Added.
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const):

Canonical link: https://commits.webkit.org/279115@main
